pdg/src/info.rs

```diff
@@ -0,0 +1,515 @@
+        flows_to_pos_offset: node_does_pos_offset(n).then(|| n_id),
+        flows_to_neg_offset: node_does_neg_offset(n).then(|| n_id)
+    }
+}
+
+fn create_flow_info (g: &Graph) -> HashMap<NodeId,GraphTraverseInfo> {
+    let mut f = HashMap::from_iter(
+        g.nodes.iter_enumerated()
+        .map(|(idx,node)| (idx,init_traverse_info(idx,node)))
+    );
+    for (n_id,node) in g.nodes.iter_enumerated().rev(){ 
+        let cur : GraphTraverseInfo = *(f.get(&n_id).unwrap());
+        if let Some(p_id) = node.source {
+            let parent = f.get_mut(&p_id).unwrap();
+            parent.last_descendent = cmp::max(cur.last_descendent,parent.last_descendent);
+            parent.flows_to_load = parent.flows_to_load.or(cur.flows_to_load);
+            parent.flows_to_store = parent.flows_to_store.or(cur.flows_to_store);
+            parent.flows_to_pos_offset = parent.flows_to_pos_offset.or(cur.flows_to_pos_offset);
+            parent.flows_to_neg_offset = parent.flows_to_neg_offset.or(cur.flows_to_neg_offset);
+        }
+    }
+    f
+}
+
+fn collect_children (g: &Graph) -> HashMap<NodeId,Vec<NodeId>> {
+    let mut m = HashMap::new();
+    for (par,chi) in g.nodes.iter_enumerated().filter_map(|(idx,node)| Some((node.source?,idx))){
+        m.entry(par).or_insert_with(Vec::new).push(chi)
+    };
+    for (par,chi) in g.nodes.iter_enumerated(){
+        m.try_insert(par,Vec::new());
+    };
+    m
+}
+
+fn partition_into_alias_sets (g: &Graph) -> HashMap<(NodeId,Vec<Field>),Vec<NodeId>> {
+    let mut store_seen : HashMap<NodeId,(NodeId,Vec<Field>)> = HashMap::new();
+    let mut store_roots : HashMap<(NodeId,Vec<Field>),Vec<NodeId>> = HashMap::new();
+
+    for n_id in g.nodes.indices(){
+        let node = g.nodes.get(n_id).unwrap();
+        match node.source {
+            None => {
+                store_roots.insert((n_id,Vec::new()),vec![n_id]);
+                store_seen.insert(n_id,(n_id,Vec::new()))
+            },
+            Some(par_id) => {
+                let (parent_root,parent_fields) = store_seen.get(&par_id).unwrap().clone();
+                match node.kind {
+                    NodeKind::Field(f) => {
+                        let mut cp = parent_fields;
+                        cp.push(f);
+                        store_roots.entry((parent_root,cp.clone())).or_insert_with(Vec::new).push(n_id);
+                        store_seen.insert(n_id,(parent_root, cp))
+                    },
+                    _ => store_seen.insert(n_id,(parent_root, parent_fields)),
+                }
+            }
+        };
+    };
+    store_roots
+}
+
+fn check_sibling_conflict(siblings: &mut Vec<NodeId>, flow_info: &HashMap<NodeId,GraphTraverseInfo>, conflict_result: &mut HashMap<NodeId,NodeId>){
+    let mut max_desc : NodeId = *siblings.get(0).unwrap();
+    let mut max_desc_parent : NodeId = *siblings.get(0).unwrap();
+    for id in siblings {
+        if *id < max_desc {
```

Comment:
it seems like this assumes some invariant for `siblings`. It would be good to document what that is, add an assertion, or encode it in the type of `siblings` so that the user doesn't accidentally call this function with input that's semantically incorrect.

Reply:
Okay, will add that in.